seqr/views/apis/auth_api.py

```diff
@@ -35,6 +42,111 @@ def login_view(request):
+
+  authorization_url, state = flow.authorization_url(
+      # Enable offline access so that you can refresh an access token without
+      # re-prompting the user for permission. Recommended for web server apps.
+      access_type='offline',
+      # Enable incremental authorization. Recommended as a best practice.
+      include_granted_scopes='true')
+
+  # Store the state so the callback can verify the auth server response.
+  request.session['state'] = state
+
+  # Return the url to be sent for the Google Auth server by the the front end
+  return create_json_response({'data': authorization_url})
+
+@csrf_exempt
+def login_oauth2callback(request):
+    """
+    Callback function will be called after the OAuth2 server passes the authentication
+
+    In the previous function 'login_google', a redirect url is generated with the information of Google client ID and
+    the front end will be redirect to the url to do authentication with Google auth server. Once the authentication has
+    passed, the result of the authentication will be sent to the web server through this callback function.
+    :param request: includes the authorized code
+    :return: status of authorization server and resource server (which is AnVIL server) responses
+    """
+
+    # Specify the state when creating the flow in the callback so that it can
+    # verified in the authorization server response.
+    state = request.session['state']
+
+    flow = google_auth_oauthlib.flow.Flow.from_client_config(
+        GOOGLE_AUTH_CLIENT_CONFIG, scopes = scopes, state = state)
+    flow.redirect_uri = GOOGLE_AUTH_CLIENT_CONFIG['web']['redirect_uris'][0]
+
+    # Use the authorization server's response to fetch the OAuth 2.0 tokens.
+    authorization_response = request.body.decode('utf-8')
+    flow.fetch_token(authorization_response = authorization_response)
+
+    token = flow.credentials.id_token
+
+    try:
+        # Decode the id token (It is a JWT token actually) to get user ID info
+        idinfo = id_token.verify_oauth2_token(token, requests.Request())
+    except ValueError as ve:
+        logger.warning('Login attempt failed for verify OAuth2 token exception.')
+        return create_json_response({'message': str(ve)}, status=401, reason=str(ve))
+
+    credentials = Credentials(**credentials_to_dict(flow.credentials))
+    session = AnvilSession(credentials = credentials, scopes = scopes)
+    try:
+        # Todo: use the anvil profile for the user name instead of using the name from the User model
+        _ = session.get_anvil_profile()
+    except Exception as ee:
+        logger.warning('User {} attempted logging in without registering with AnVIL.'.format(idinfo['email']))
+        return create_json_response({'message': 'Google account {} has\'t been registered on AnVIL yet. '
+            'Please open https://anvil.terra.bio to sign in with Google and register the account.'.format(idinfo['email'])},
```

Comment:
Can you make the link an actual link (in an `<a>` tag)

Reply:
Yes. Will change it.